DESIGN.md

```diff
@@ -0,0 +1,58 @@
+# Module Design
+
+## Classes
+
+There are 4 main public classes and other can be added if they need to be managed separately than the existing classes:
+
+* sensu - Resources shared by all other public classes such as agent and backend
+* sensu::agent - Sensu Agent
+* sensu::backend - Sensu Backend
+* sensu::plugins - Sensu plugins
+
+Subclasses to the above hold resources for that resources to keep the logic in the main public class shorter. So `sensu::backend::default_resources` is a private classes that adds functionality to `sensu::backend`. There are also some shared classes like `sensu::ssl` that are private and have resources shared by other public classes.
+
+If a parameter is used by both agent and backend, it belongs in `sensu` class. All other parameters should be added to the appropriate public class.
+
+## Types/Provides
+
+### Sensuctl types
+
+The types to `sensuctl` resources are designed to closely match the resource specifications from Sensu Go. For example the Sensu Go `check` resource property of `interval` would map directly to the `sensu_check` type property of `interval`.
+
+Keys in the Sensu Go specification under `spec` should have a 1-to-1 mapping with Puppet type properties.
+
+The exception to the 1-to-1 mapping is Sensu Go resource `metadata`. The values for `metadata` keys of `name`, `namespace`, `labels`, and `annotations` are pulled out into Puppet type properties. The metadata exception is to properly handle resource names and namespaces and relationships as well as the composite names used by namespace capable resources.
+
+### All other types/providers
+
+Other types and providers are added where possible to avoid the use of Exec resources in the manifest code. One example is `sensu_plugin` which manages Sensu Go plugins that doesn't use `sensuctl`.
+
+The use of Exec is acceptable for simple cases like adding the license file where the Exec can be triggered when a file changes.
+
+## Unit tests
+
+The unit tests are designed to check Puppet catalog behavior without ensuring the functionality on a running system.
+
+Test locations:
+
+* `spec/classes` - Theses for each class in `manifests` directory.
+* `spec/unit/*_spec.rb` - These are tests for the types defined in `lib/puppet/types`
+* `spec/unit/provider/**/*_spec.rb` - These are tests for the providers defined in `lib/puppet/providers`
```

Comment:
should document the rest of the tests

Reply:
What other tests are there? Some of this is redundant to documentation from rspec-puppet as we are doing nothing custom for test placement except for things under `spec/unit`. I've added type aliases location but not sure what other tests exist.

Reply:
Never mind about other tests, didn't scroll far enough down.